src/snmalloc/mem/metadata.h

```diff
@@ -500,6 +500,58 @@ namespace snmalloc
       return (--needed()) == 0;
     }
 
+    class ReturnObjectsIterator
+    {
+      uint16_t _batch;
+      FrontendSlabMetadata* _meta;
+
+      static_assert(sizeof(_batch) * 8 > MAX_CAPACITY_BITS);
+
+    public:
+      ReturnObjectsIterator(uint16_t n, FrontendSlabMetadata* m)
+      : _batch(n), _meta(m)
+      {}
+
+      template<bool first>
+      SNMALLOC_FAST_PATH bool step()
+      {
+        if (SNMALLOC_LIKELY(_batch < _meta->needed()))
+        {
+          // Will not hit threshold for state transition
+          _meta->needed() -= _batch;
+          return false;
+        }
+
+        /*
+         * If we've already taken a step and there are no more unreturned
+         * objects, then we don't need to take another step.  It can happen
+         * that we get here because _meta->needed() is still 0, tho', so
+         * bail.
+         */
+        if (!first && _batch == 0)
+          return false;
```

Comment:
> I'm unsure what I could have done to make Windows unhappy, but that suggest I've broken something subtle. :(
> 
> I no longer have a Windows machine on which to test things; can I ask you to have a look, @mjp41?

I haven't tried on Windows yet.  But just inspecting the CI, I am wondering if it is valid to read needed after we have performed the last meta data update.  Could another thread have reused the meta-data by this point or turned the page behind the meta-data off?  Perhaps, we need to use the batch size first in the cases where we can.  Which is effectively what your original code did as it would only look another time if it was non-zero.
```suggestion
        SNMALLOC_ASSERT(first && (batch != 0));

        /*
         * If we've already taken a step and there are no more unreturned
         * objects, then we don't need to take another step.  It can happen
         * that we get here because _meta->needed() is still 0, tho', so
         * bail.
         */
        if (!first && _batch == 0)
          return false;
        
        if (SNMALLOC_LIKELY(_batch < _meta->needed()))
        {
          // Will not hit threshold for state transition
          _meta->needed() -= _batch;
          return false;
        }
```

Reply:
FYI, updated the suggestion above to fix my failure to do basic logic.

Reply:
Ah, that's a good point.  Lemme apply that fix.

Reply:
I don't think that `SNMALLOC_ASSERT` is right, though.  The thing the new condition is guarding against is precisely the case that we have taken the first step (and enough subsequent steps) to have returned all objects.

Reply:
Sorry, I'm a bit confused why the first one would be empty?  Perhaps, applying de-Morgan to the assert would be helpful:
```C++
   SNMALLOC_ASSERT(!(first && batch == 0));
```

Reply:
Indeed, I just couldn't read.  I'll add the assert.